harness/tests/experiment/pytorch/test_reducer.py

```diff
@@ -12,3 +23,96 @@ def test_reducer() -> None:
+    dataset_size = 47
+
+    def do_parallel(fn: Callable) -> List:
+        """
+        Run the same function on one-thread-per-rank, assert there were no exceptions, and return
+        the results from each rank.
+        """
+        results = [None] * size  # type: List
+        errors = [None] * size  # type: List
+        threads = []
+
+        for cross_rank, local_rank in itertools.product(range(cross_size), range(local_size)):
+            rank = cross_rank * local_size + local_rank
+
+            def _fn(rank: int, cross_rank: int, local_rank: int) -> None:
+                try:
+                    results[rank] = fn(rank, cross_rank, local_rank)
+                except Exception:
+                    errors[rank] = traceback.format_exc()
+                    raise
+
+            threads.append(threading.Thread(target=_fn, args=(rank, cross_rank, local_rank)))
+
+        for thread in reversed(threads):
+            thread.start()
+
+        for thread in threads:
+            thread.join()
+
+        assert errors == [None] * size, "not all threads exited without error"
+
+        return results
+
+    def make_reducer_context(rank: int, cross_rank: int, local_rank: int) -> DummyTrial:
+        distributed_context = _core.DistributedContext(
+            rank=cross_rank * local_size + local_rank,
+            size=cross_size * local_size,
+            local_rank=local_rank,
+            local_size=local_size,
+            cross_rank=cross_rank,
+            cross_size=cross_size,
+            chief_ip="localhost",
+            force_tcp=False,
+        )
+        reducer_context = _PyTorchReducerContext(distributed_context._zmq_allgather)
+        # reducer_context.wrap_reducer(lambda x: x, "dummy")
+        wrapped_reducer = reducer_context.wrap_reducer(dummy_reducer)
+        return DummyTrial(distributed_context, reducer_context, wrapped_reducer)
+
+    trials = do_parallel(make_reducer_context)
+
+    def get_batch_list(
+        rank: int, batch_size: int, num_workers: int, seq: List[int]
+    ) -> List[List[int]]:
+        total_batches = (len(seq) + (batch_size - 1)) // batch_size
+        my_batch_indices = [i for i in range(total_batches) if i % num_workers == rank]
+        all_batches = [
+            seq[batch_size * k : min(batch_size * k + batch_size, len(seq))]
+            for k in range(total_batches)
+        ]
+        return [b for i, b in enumerate(all_batches) if i in my_batch_indices]
+
+    observations = list(range(dataset_size))
+    for rank, trial in enumerate(trials):
+        for batch in get_batch_list(rank, 2, len(trials), observations):
+            trial.wrapped_reducer.update(batch)
+
+    results = do_parallel(lambda rank, _, __: trials[rank].reducer_context.reduce_metrics(False))
+    logger.debug(results)
+
+    # Close all distributed contexts
+    for trial in trials:
+        trial.distributed_context.close()
+
+    assert results[0]["sum"] == dataset_size * (dataset_size - 1) // 2
```

Comment:
maybe test all of the results?  Not sure it's critical though.

Reply:
Not sure we care. I mean, this would mostly be testing pre-existing functionality (the fact that chief and all workers see the same thing after `_zmq_allgather`). I guess it's not bad to have net new coverage.